tfhe/src/high_level_api/tag.rs

```diff
@@ -0,0 +1,331 @@
+use crate::high_level_api::backward_compatibility::tag::{SmallVecVersions, TagVersions};
+use tfhe_versionable::Versionize;
+
+/// Simple short optimized vec, where if the data is small enough
+/// (<= std::mem::size_of::<Vec<u8>>() - 1) the data will be stored on the stack
+///
+/// Once a true heap allocated Vec was needed, it won't be deallocated in favor
+/// of stack data.
+#[derive(Clone, Debug, Versionize)]
```

Comment:
When someone call `small_vec.versionize().serialize()`, the custom impl of `Serialize` for SmallVec won't be called because the call to `serialize` will be made on type `SmallVec::Versioned`.

However, if the idea is to convert this to a plain slice before serialization, this struct does not need to be versioned in itself because it will never be stored itself (it is only a runtime type). So I believe you can directly store the array without versioning info.
This can be done using the `from/into` attribute of the macro (https://github.com/zama-ai/tfhe-rs/blob/main/utils/tfhe-versionable/examples/convert.rs) or using a manual impl of `Versionize/VersionizeOwned/Unversionize` (for example with `type Versioned<'vers> = &'vers [u8]`).

Reply:
But we would still need the custom serialize impl as the `versionize()` thing is opt-in rather than opt-out (and even if it were opt-out we would still need it)

Reply:
yes ! The goal is not to remove the Serialize impl but to have a consistent serialization behavior. If you want to factorize the conversion code between serialize and versionize you can use from/into (it's [also supported by serde](https://serde.rs/container-attrs.html#from)), but to be fair I'm not sure it will work with versionize in your case.

Reply:
One small thing is that serde's from/into would require to clone the data that is inside the small vec into a new vec so it can be serialized, which is a bit less efficient than the custom impl where i can `serialize_bytes(self.data())`

Also, that means the `Tag` also do not need to be versionned

Reply:
You're right the custom impl should be more efficient !

I think the Tag needs to be versioned because it is not guaranteed to always be represented as a wrapper around an array-like type. However the `SmallVec` is an optimization that only makes sense in a running application and will always be equivalent to an array in the serialization data model.

Reply:
yes it makes sense